### PR TITLE
ENH: add constant and white noise kernel, fix docs and tests

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
I have added the constant kernel, fixed some tests and also fixed some docs in GP module. Please feel free to leave a review on the work done! @fonnesbeck @aloctavodia
- [x] add `diag` parameter to the covariance function.
- [x] ~add diag method in all the cov functions to evaluate just the diagonal instead of computing the whole cov matrix.~
- [x] add `active_dims` to compute with hand-picked `feature_ndims` columns.
- [x] optimize the combination class of covariance functions.
- [x] ~start putting together a notebook on covariance and mean functions.~
- [x] refactor and add tests for new features.
- [x] fix failing docs in the submodule `gp`

### CODE REVIEW

File: pymc4/gp/cov.py [hunk 1]
```diff
@@ -149,9 +189,11 @@ class CovarianceProd(Combination):
     Parameters
     ----------
     feature_ndims : int
-        number of rightmost dims to include in kernel computation
+        The number of rightmost dimensions to be absorbed during
```
Comment (aloctavodia) [inline pymc4/gp/cov.py:192]: nitpicking. The line-length limit is 100.

File: pymc4/gp/mean.py [hunk 2]
```diff
@@ -32,9 +43,9 @@ class MeanAdd(Mean):
 
     Parameters
     ----------
-    mean1 : {callable, pm.Mean}
+    mean1 : {callable, pymc4.gp.Mean}
```
Comment (aloctavodia) [inline pymc4/gp/mean.py:46]: I think is ok to use `pm`

File: pymc4/gp/cov.py [hunk 3]
```diff
@@ -8,42 +8,58 @@
...
+from .constant import _Constant
+from .white_noise import _WhiteNoise
+from .util import ArrayLike, TfTensor, _inherit_docs
 
 
 __all__ = [
-    #     "Constant",
-    #     "WhiteNoise",
+    "Constant",
+    # "WhiteNoise",
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply pymc4/gp/cov.py]: done!

File: pymc4/gp/cov.py [hunk 4]
```diff
@@ -8,42 +8,58 @@
 import numpy as np
 import tensorflow_probability as tfp
 
-from .util import ArrayLike, TfTensor
+from .constant import _Constant
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply pymc4/gp/cov.py]: done!

File: pymc4/gp/mean.py (hunk 2 above)
Comment (tirthasheshpatel) [reply pymc4/gp/mean.py:46]: I will check that out. I think it doesn't affect the doc tests by pydocstyle. So any name would work.

File: pymc4/gp/cov.py (hunk 1 above)
Comment (tirthasheshpatel) [reply pymc4/gp/cov.py:205]: true! I am somehow always a little scared to exceed the limit. will fix that all over the module

**Comment by tirthasheshpatel** [comment]
failures don't seem to be caused by the changes in this PR. Somehow azure is failing to install `black`! Can someone please trigger the builds again?

**Comment by tirthasheshpatel** [comment]
I have covered almost everything and added some features for arbitrarily shaped tensors that weren't in pymc3.
 - `active_dims`: This is a list or a list of lists with each entry indicating the dimensions of each of the `feature_ndims` columns to be used during computation of the cov matrix. I have written some tests showing many [examples and their behaviour here](https://github.com/pymc-devs/pymc4/pull/272/files#diff-0cdd7a01c76ee7c96ad873e46b9db56cR180).
 - `scale_diag`: Performs ard. If `ARD=False`, this argument is ignored.
 - `_kernel` is a private module containing kernels not implemented in tfp.
 - In `_diag` method of covariance functions, I have currently kept an argument `to_dense`. This parameter asks weather to return result as a diagonal vector or a square matrix. This may be useful if I am able to find an algorithm to work with diagonal vector in GP models instead of the whole matrix. It would be efficient in both time and space. I would love to hear others thought on this.

Will appreciate reviews now to address any mistakes or nitpicks. Should I add a notebook explaining mean and covariance functions or should I leave it to some other PR?

File: pymc4/gp/_kernel.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
+"""Constant Kernel and WhiteNoise Kernel."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow.compat.v2 as tf  # pylint: disable=import-error
```
Comment (fonnesbeck) [inline pymc4/gp/_kernel.py:7]: Do we still need to use `compat.v2`?

File: pymc4/gp/mean.py [hunk 6]
```diff
@@ -28,76 +49,86 @@ def __mul__(self, mean2):
...
     feature_ndims : int, optional
-        number of rightmost dims to include in mean computation
+        number of rightmost dims to include in mean computation. (default=1)
     """
 
+    @_inherit_docs(Mean.__call__)
     def __call__(self, X: ArrayLike) -> TfTensor:
-        X = tf.convert_to_tensor(X)
-        return tf.zeros(X.shape[: -self.feature_ndims])
+        dtyp = dtype_util.common_dtype([X])
```
Comment (fonnesbeck) [inline pymc4/gp/mean.py:109]: You can call it `dtype`, as its not a reserved word.

File: pymc4/gp/_kernel.py (hunk 5 above)
Comment (tirthasheshpatel) [reply pymc4/gp/_kernel.py]: Oh sorry! Was using that for consistency with tfp's convention. I should be changed now.

File: pymc4/gp/mean.py (hunk 6 above)
Comment (tirthasheshpatel) [reply pymc4/gp/mean.py]: changed!

**Comment by fonnesbeck** [comment]
Thanks @tirthasheshpatel !

**Pull request merged**